consensus/src/experimental/tests/commit_phase_tests.rs

```diff
@@ -0,0 +1,383 @@
+    let network = NetworkSender::new(author, network_sender, self_sender, validators);
+
+    let (commit_result_sender, commit_result_receiver) =
+        channel::new_test::<(Vec<Block>, LedgerInfoWithSignatures)>(channel_size);
+    let state_computer = Arc::new(OrderingStateComputer::new(commit_result_sender));
+
+    let mut safety_rules = MetricsSafetyRules::new(safety_rules_manager.client(), storage);
+    safety_rules.perform_initialize().unwrap();
+
+    let safety_rules_container = Arc::new(Mutex::new(safety_rules));
+
+    // setting up channels
+    let (sender_comm, receiver_comm) =
+        channel::new_test::<(Vec<ExecutedBlock>, LedgerInfoWithSignatures)>(channel_size);
+
+    let (sender_c_msg, receiver_c_msg) = channel::new_test::<ConsensusMsg>(channel_size);
+
+    let commit_phase = CommitPhase::new(
+        receiver_comm,
+        state_computer.clone(),
+        receiver_c_msg,
+        epoch_state.verifier.clone(),
+        safety_rules_container.clone(),
+        author,
+        back_pressure,
+        network,
+    );
+
+    runtime.spawn(commit_phase.start());
+    (
+        sender_comm,            // channel to pass executed blocks into the commit phase
+        sender_c_msg,           // channel to pass commit messages into the commit phase
+        commit_result_receiver, // channel to receive commit result from the commit phase
+        self_receiver,          // channel to receive message from the commit phase itself
+        safety_rules_container,
+        signers,
+        state_computer,
+        epoch_state.verifier,
+        runtime,
+    )
+}
+
+fn prepare_commit_arguments(
+    signer: &ValidatorSigner,
+    state_computer: Arc<OrderingStateComputer>,
+) -> (Vec<ExecutedBlock>, LedgerInfoWithSignatures) {
+    let genesis_qc = certificate_for_genesis();
+    let block = Block::new_proposal(vec![], 1, 1, genesis_qc, signer);
+    let compute_result = state_computer
+        .compute(&block, *ACCUMULATOR_PLACEHOLDER_HASH)
+        .unwrap();
+
+    let li = LedgerInfo::new(
+        block.gen_block_info(*ACCUMULATOR_PLACEHOLDER_HASH, 0, None),
+        *ACCUMULATOR_PLACEHOLDER_HASH,
+    );
+
+    let mut li_sig = LedgerInfoWithSignatures::new(
+        li.clone(),
+        BTreeMap::<AccountAddress, Ed25519Signature>::new(),
+    );
+
+    li_sig.add_signature(signer.author(), signer.sign(&li));
+
+    (vec![ExecutedBlock::new(block, compute_result)], li_sig)
+}
+
+/// happy path test
+fn test_happy_path() {
+    let (
+        mut sender_comm,
+        mut sender_c_msg,
+        mut commit_result_receiver,
+        mut self_receiver,
+        _safety_rules_container,
+        signers,
+        state_computer,
+        _validator,
+        _runtime,
+    ) = prepare_commit_phase();
+
+    // send good commit arguments
+    block_on(sender_comm.send(prepare_commit_arguments(&signers[0], state_computer))).ok();
+
+    match block_on(self_receiver.select_next_some()) {
```

Comment:
use next() instead, only use select_ inside a select! macro, also can use a  async closure + timed_block_on to avoid multiple block_on

Reply:
resolved